runtime/src/lib.rs

```diff
@@ -549,26 +573,60 @@ impl additional_traits::DelegatedDispatchVerifier for Runtime {
 	const DOMAIN: &'static str = "cennznet";
 
 	fn verify_dispatch(doughnut: &Self::Doughnut, module: &str, method: &str) -> Result<(), &'static str> {
-		let mut domain = doughnut
-			.get_domain(Self::DOMAIN)
-			.ok_or("CENNZnut does not grant permission for cennznet domain")?;
-		let cennznut: CENNZnut = Decode::decode(&mut domain).map_err(|_| "Bad CENNZnut encoding")?;
+		let cennznut: CENNZnut = Self::get_cennznut(doughnut)?;
 
 		// Extract Module name from <prefix>-<Module_name>
-		let module_offset = module.find('-').ok_or("error during module name segmentation")? + 1;
+		let module_offset = module
+			.find('-')
+			.ok_or("CENNZnut does not grant permission for module")?
+			+ 1;
 		if module_offset <= 1 || module_offset >= module.len() {
 			return Err("error during module name segmentation");
 		}
-		match cennznut.validate(&module[module_offset..], method, &[]) {
+		match cennznut.validate_runtime_call(&module[module_offset..], method, &[]) {
 			Ok(r) => Ok(r),
 			Err(ValidationErr::ConstraintsInterpretation) => Err("error while interpreting constraints"),
-			Err(ValidationErr::NoPermission(Domain::Method)) => Err("CENNZnut does not grant permission for method"),
-			Err(ValidationErr::NoPermission(Domain::Module)) => Err("CENNZnut does not grant permission for module"),
-			Err(ValidationErr::NoPermission(Domain::MethodArguments)) => {
+			Err(ValidationErr::NoPermission(RuntimeDomain::Method)) => {
+				Err("CENNZnut does not grant permission for method")
+			}
+			Err(ValidationErr::NoPermission(RuntimeDomain::Module)) => {
+				Err("CENNZnut does not grant permission for module")
+			}
+			Err(ValidationErr::NoPermission(RuntimeDomain::MethodArguments)) => {
 				Err("CENNZnut does not grant permission for method arguments")
 			}
 		}
 	}
+
+	/// Verify that the contract being called is permissioned on the doughnut
+	/// This is used in two cases:
+	/// * Case 1 - The `holder` of a doughnut wants to invoke a smart contract as the `issuer`
+	/// * Case 2 - The `issuer` wants to execute a contract but only with the permissions avaliable
+	///            in the doughnut being sent. (not implemented yet)
+	fn verify_runtime_to_contract_call(
+		caller: &Self::AccountId,
+		doughnut: &Self::Doughnut,
+		contract_addr: &Self::AccountId,
+	) -> Result<(), &'static str> {
+		if caller.clone() == doughnut.issuer() {
```

Comment:
Can we make this a `debug_assert!` and with the correct tests should be ok 💭

Reply:
Yes we can (and have!)